google/auth/credentials.py

```diff
@@ -0,0 +1,209 @@
+    construction. Some classes will provide mechanisms to copy the credentials
+    with modifications such as :meth:`ScopedCredentials.with_scopes`.
+    """
+    def __init__(self):
+        self.token = None
+        """str: The bearer token that can be used in HTTP headers to make
+        authenticated requests."""
+        self.expiry = None
+        """Optional[datetime]: When the token expires and is no longer valid.
+        If this is None, the token is assumed to never expire."""
+
+    @property
+    def valid(self):
+        """Checks the validity of the credentials.
+
+        This is True if the credentials have a :attr:`token` and the token
+        is not :attr:`expired`.
+        """
+        return self.token is not None and not self.expired
+
+    @property
+    def expired(self):
+        """Checks if the credentials are expired.
+
+        Note that credentials can be invalid but not expired becaue Credentials
+        with :attr:`expiry` set to None is considered to never expire.
+        """
+        now = _helpers.utcnow()
+        if self.expiry is None or self.expiry > now:
+            return False
+        else:
+            return True
+
+    @abc.abstractmethod
+    def refresh(self, request):
+        """Refreshes the access token.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+
+        Raises:
+            google.auth.exceptions.RefreshError: If the credentials could
+                not be refreshed.
+        """
+        # pylint: disable=missing-raises-doc
+        # (pylint doesn't recognize that this is abstract)
+        raise NotImplementedError('Refresh must be implemented')
+
+    def apply(self, headers, token=None):
+        """Apply the token to the authentication header.
+
+        Args:
+            headers (Mapping): The HTTP request headers.
+            token (Optional[str]): If specified, overrides the current access
+                token.
+        """
+        headers[b'authorization'] = 'Bearer {}'.format(
+            _helpers.from_bytes(token or self.token))
+
+    def before_request(self, request, method, url, headers):
+        """Performs credential-specific before request logic.
+
+        Refreshes the credentials if necessary, then calls :meth:`apply` to
+        apply the token to the authentication header.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+            method (str): The request's HTTP method.
+            url (str): The request's URI.
+            headers (Mapping): The request's headers.
+        """
+        # pylint: disable=unused-argument
+        # (Subclasses may use these arguments to ascertain information about
+        # the http request.)
+        if not self.valid:
+            self.refresh(request)
+        self.apply(headers)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ScopedCredentials(object):
```

Comment:
I'm not following the diamond inheritance problem worry. If all three interfaces contain no concrete elements (not even private ones), there's no diamond inheritance problem, right? It's not like what would be the two subinterfaces have conflicting method names.

Yes, I am mostly fine with this as long as the concrete elements are dropped from the interfaces.

For common behavior, implement it once in one (private) place and delegate to it from the n places where it is exposed. Have it directly tested by an access-to-implementation test and have it indirectly tested by the API-surface-level tests that test each place where it is exposed. Yes, these extra connections are [forwarding methods](https://en.wikipedia.org/wiki/Composition_over_inheritance#Drawbacks) that in isolation would be considered redundant cruft but in the context of _allowing you to separate the way your system is presented from the way it is implemented_ are a price well worth paying. Absolutely don't repeat the executable statements that form the actual useful logic of the library; that would be bloat as well as inviting bugs and inconsistency to creep in over time.

It may be the case that you choose for the private implementation of the library classes very similar to those originally drafted for the library's public API. Should that come to pass, I won't raise much of a fuss since they'll be all private implementation.